src/Microsoft.Diagnostics.Monitoring.EventPipe/DiagnosticsEventPipeProcessor.cs

```diff
@@ -53,10 +57,18 @@ public class DiagnosticsEventPipeProcessor : IAsyncDisposable
             _metricIntervalSeconds = metricIntervalSeconds;
             _logsLevel = logsLevel;
             _processInfoCallback = processInfoCallback;
+            _userConfig = configuration;
+            _streamAvailable = streamAvailable;
+            _processInfoCallback = processInfoCallback;
+            _counterFilter = metricFilter;
+
+            _sessionStarted = new TaskCompletionSource<bool>(TaskCreationOptions.RunContinuationsAsynchronously);
         }
 
         public async Task Process(DiagnosticsClient client, int pid, TimeSpan duration, CancellationToken token)
         {
+            //No need to guard against reentrancy here, since the calling pipeline does this already.
+            IDisposable registration = token.Register(() => _sessionStarted.TrySetCanceled());
```

Comment:
This can be simplified by moving it to inside the StartNew lambda and use `using var _ = token.Register(...);` instead of manually managing the disposal.

Reply:
I'm manually disposing elsewhere too, I will leave this as is for now.